File: _includes/code/python/howto.configure.rbac.permissions.py
import weaviate
from weaviate import WeaviateClient
from weaviate.classes.init import Auth

client = weaviate.connect_to_local(
    # Use custom port defined in tests/docker-compose-rbac.yml (without showing the user)
    port=8580,
    grpc_port=50551,
    auth_credentials=Auth.api_key("root-user-key"),
)


def reset_user(user: str, client: WeaviateClient):
    # Clean slate
    current_roles = client.users.get_assigned_roles(user)  # check if user exists
    for k in current_roles.keys():
        client.users.revoke_roles(user_id=user, role_names=k)  # revoke all roles


# =================================================================
# =============== EXAMPLE: READ + WRITE PERMISSIONS
# =================================================================

# Clean slate
reset_user("custom-user", client=client)
client.roles.delete("rw_role")  # delete if exists

# START ReadWritePermissionDefinition
from weaviate.classes.rbac import Permissions

# Define permissions (example confers read+write rights to collections starting with "TargetCollection")
permissions = [
    # Collection level permissions
    Permissions.collections(
        collection="TargetCollection*",
        create_collection=True,  # Allow creating new collections
        read_config=True,  # Allow reading collection info/metadata
        update_config=True,  # Allow updating collection configuration, i.e. update schema properties, when inserting data with new properties
        delete_collection=True,  # Allow deleting collections
    ),
    # Collection data level permissions
    Permissions.data(
        collection="TargetCollection*",
        create=True,  # Allow data inserts
        read=True,  # Allow query and fetch operations
        update=True,  # Allow data updates
        delete=False,  # Allow data deletes
    ),
    Permissions.backup(collection="TargetCollection*", manage=True),
    Permissions.Nodes.verbose(collection="TargetCollection*", read=True),
    Permissions.cluster(read=True),
]

# Create a new role
client.roles.create(role_name="rw_role", permissions=permissions)
# END ReadWritePermissionDefinition
# START ReadWritePermissionAssignment
# Assign the role to a user
client.users.assign_roles(user_id="custom-user", role_names=["rw_role"])
# END ReadWritePermissionAssignment

# ===== TEST ===== basic checks to see if the role was created
user_permissions = client.users.get_assigned_roles("custom-user")

assert "rw_role" in user_permissions.keys()
assert (
    user_permissions["rw_role"].collections_permissions[0].collection
    == "TargetCollection*"
)
assert user_permissions["rw_role"].name == "rw_role"

# =================================================================
# =============== EXAMPLE: VIEWER PERMISSIONS
# =================================================================

# Clean slate
client.roles.delete("viewer_role")  # delete if exists

# START ViewerPermissionDefinition
from weaviate.classes.rbac import Permissions

# Define permissions (example confers viewer rights to collections starting with "TargetCollection")
permissions = [
    Permissions.collections(
        collection="TargetCollection*",
        read_config=True,
    ),
    Permissions.data(collection="TargetCollection*", read=True),
]

# Create a new role
client.roles.create(role_name="viewer_role", permissions=permissions)
# END ViewerPermissionDefinition
# START ViewerPermissionAssignment
# Assign the role to a user
client.users.assign_roles(user_id="custom-user", role_names="viewer_role")
# END ViewerPermissionAssignment

# =================================================================
# =============== EXAMPLE: VIEWER PERMISSIONS
# =================================================================

# Clean slate
client.roles.delete("tenant_manager")

# START MTPermissionsExample
from weaviate.classes.rbac import Permissions

permissions = [
    Permissions.tenants(
        collection="TargetCollection*",  # Applies to all collections starting with "TargetCollection"
        tenant="TargetTenant*",  # Applies to all tenants starting with "TargetTenant"
        create=True,  # Allow creating new tenants
        read=True,  # Allow reading tenant info/metadata
        update=True,  # Allow updating tenant states
        delete=True,  # Allow deleting tenants
    ),
    Permissions.data(
        collection="TargetCollection*",  # Applies to all collections starting with "TargetCollection"
        tenant="TargetTenant*",  # Applies to all tenants starting with "TargetTenant"
        create=True,  # Allow data inserts
        read=True,  # Allow query and fetch operations
        update=True,  # Allow data updates
        delete=True,  # Allow data deletes
    ),
]

# Create a new role
client.roles.create(role_name="tenant_manager", permissions=permissions)
# END MTPermissionsExample
# START MTPermissionsAssignment
# Assign the role to a user
client.users.assign_roles(user_id="custom-user", role_names="tenant_manager")
# END MTPermissionsAssignment

# ===== TEST ===== basic checks to see if the role was created
user_permissions = client.users.get_assigned_roles("custom-user")

assert "viewer_role" in user_permissions.keys()
assert (
    user_permissions["viewer_role"].collections_permissions[0].collection
    == "TargetCollection*"
)
assert user_permissions["viewer_role"].name == "viewer_role"

client.close()

File: _includes/code/typescript/howto.configure.rbac.permissions.ts
import weaviate, { WeaviateClient } from 'weaviate-client'
import assert from 'assert'

const client: WeaviateClient = await weaviate.connectToLocal({
        // Use custom port defined in tests/docker-compose-rbac.yml (without showing the user)
        port: 8580,
        grpcPort: 50551, 
        authCredentials: new weaviate.ApiKey("user-a-key")
})


async function resetUser(user: string, client: WeaviateClient) {
    // Clean slate
    const currentRoles = await client.users.getAssignedRoles(user)  // check if user exists
    for await (const [role, value] of Object.entries(currentRoles)) {
        await client.users.revokeRoles(role, user)  // revoke all roles
    }
}
// =================================================================
// =============== EXAMPLE: READ + WRITE PERMISSIONS
// =================================================================

// Clean slate
resetUser("user-b", client)
await client.roles.delete("rw_role")  // delete if exists

// START ReadWritePermissionDefinition // START MTPermissionsExample
const { permissions } = weaviate
// END ReadWritePermissionDefinition // END MTPermissionsExample

// START ReadWritePermissionDefinition

// Define permissions (example confers read+write rights to collections starting with "TargetCollection")
const allPermissions = [
    // Collection level permissions
    permissions.collections({
        collection: "TargetCollection*",
        create_collection:  true,  // Allow creating new collections
        read_config: true,  // Allow reading collection info/metadata
        update_config: true,  // Allow updating collection configuration, i.e. update schema properties, when inserting data with new properties
        delete_collection: true,  // Allow deleting collections
   } ),
    // Collection data level permissions
    permissions.data({
        collection: "TargetCollection*",
        create: true,  // Allow data inserts
        read: true,  // Allow query and fetch operations
        update: true,  // Allow data updates
        delete: false,  // Allow data deletes
    }),
    permissions.backup({
        collection:"TargetCollection*", 
        manage: true
    }),
    permissions.nodes.verbose({
        collection: "TargetCollection*", 
        read: true
    }),
    permissions.cluster({
        read: true
    }),
]

// Create a new role
await client.roles.create("rw_role", allPermissions)
// END ReadWritePermissionDefinition

// START ReadWritePermissionAssignment
// Assign the role to a user
await client.users.assignRoles(["rw_role"], "user-b",)
// END ReadWritePermissionAssignment

// ===== TEST ===== basic checks to see if the role was created
const userPermissions = await client.users.getAssignedRoles("user-b")

assert.equal("rw_role", Object.keys(userPermissions))
assert.equal(userPermissions["rw_role"].collectionsPermissions[0].collection, "TargetCollection*")
assert.equal(userPermissions["rw_role"].name, "rw_role")

// =================================================================
// =============== EXAMPLE: VIEWER PERMISSIONS
// =================================================================

// Clean slate
await client.roles.delete("viewer_role")  // delete if exists

// START ViewerPermissionDefinition

// Define permissions (example confers viewer rights to collections starting with "TargetCollection")
const newPermissions = [
    permissions.collections({
        collection: "TargetCollection*",
        read_config: true,
    }),
    permissions.data({
        collection: "TargetCollection*", 
        read: true}),
]

// Create a new role
await client.roles.create("viewer_role", newPermissions)
// END ViewerPermissionDefinition

// START ViewerPermissionAssignment
// Assign the role to a user
await client.users.assignRoles("user-b", "viewer_role")
// END ViewerPermissionAssignment

// =================================================================
// =============== EXAMPLE: VIEWER PERMISSIONS
// =================================================================

// Clean slate
client.roles.delete("tenant_manager")

// START MTPermissionsExample

const tenantPermissions = [
    permissions.tenants({
        collection: "TargetCollection*",  // Applies to all collections starting with "TargetCollection"
        tenant: "TargetTenant*",  // Applies to all tenants starting with "TargetTenant"
        create: true,  // Allow creating new tenants
        read: true,  // Allow reading tenant info/metadata
        update: true,  // Allow updating tenant states
        delete: true,  // Allow deleting tenants
    }),
    permissions.data({
        collection: "TargetCollection*",  //  Applies to all collections starting with "TargetCollection"
        tenant: "TargetTenant*",  // Applies to all tenants starting with "TargetTenant"
        create: true,  // Allow data inserts
        read: true,  // Allow query and fetch operations
        update: true,  // Allow data updates
        delete: true,  // Allow data deletes
    })
]

// Create a new role
await client.roles.create("tenant_manager", tenantPermissions)
// END MTPermissionsExample
// START MTPermissionsAssignment
// Assign the role to a user
client.users.assignRoles("user-b", "tenant_manager")
// END MTPermissionsAssignment

// ===== TEST ===== basic checks to see if the role was created
const testUserPermissions = await client.users.getAssignedRoles("user-b")

assert.equal((Object.keys(await client.users.getAssignedRoles("user-b")).some(
    role => role == "viewer_role"
)), true)
assert.equal(
    testUserPermissions["viewer_role"].collectionsPermissions[0].collection
    ,"TargetCollection*"
)
assert.equal(testUserPermissions["viewer_role"].name, "viewer_role")

client.close()
